File: snapshot/2011/src/rac/labs/cs/Labs6.cs
#region Header
// Revit API .NET Labs
//
// Copyright (C) 2007-2009 by Autodesk, Inc.
//
// Permission to use, copy, modify, and distribute this software
// for any purpose and without fee is hereby granted, provided
// that the above copyright notice appears in all copies and
// that both that copyright notice and the limited warranty and
// restricted rights notice below appear in all supporting
// documentation.
//
// AUTODESK PROVIDES THIS PROGRAM "AS IS" AND WITH ALL FAULTS.
// AUTODESK SPECIFICALLY DISCLAIMS ANY IMPLIED WARRANTY OF
// MERCHANTABILITY OR FITNESS FOR A PARTICULAR USE.  AUTODESK, INC.
// DOES NOT WARRANT THAT THE OPERATION OF THE PROGRAM WILL BE
// UNINTERRUPTED OR ERROR FREE.
//
// Use, duplication, or disclosure by the U.S. Government is subject to
// restrictions set forth in FAR 52.227-19 (Commercial Computer
// Software - Restricted Rights) and DFAR 252.227-7013(c)(1)(ii)
// (Rights in Technical Data and Computer Software), as applicable.
#endregion // Header

#region Namespaces
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.IO;
using System.Reflection;
using System.Linq;
using WinForms = System.Windows.Forms;
using System.Windows.Media.Imaging; // for ribbon, requires references to PresentationCore and WindowsBase .NET assemblies
using Autodesk.Revit.ApplicationServices;
using Autodesk.Revit.Attributes;
using Autodesk.Revit.DB;
using Autodesk.Revit.DB.Architecture;
using Autodesk.Revit.UI;
using Autodesk.Revit.UI.Events;
using DocumentSavingEventArgs = Autodesk.Revit.DB.Events.DocumentSavingEventArgs;
#endregion // Namespaces

namespace Labs
{
  #region SampleAccessibilityCheck IExternalCommandAvailability
  public class SampleAccessibilityCheck
    : IExternalCommandAvailability
  {
    public bool IsCommandAvailable(
      UIApplication applicationData,
      CategorySet selectedCategories )
    {
      // Allow button click if there is
      // no active selection

      if( selectedCategories.IsEmpty )
        return true;

      // Allow button click if there is
      // at least one wall selected

      foreach( Category c in selectedCategories )
      {
        if( c.Id.IntegerValue
          == ( int ) BuiltInCategory.OST_Walls )
          return true;
      }
      return false;
    }
  }
  #endregion // SampleAccessibilityCheck IExternalCommandAvailability

  #region Lab6_1_HelloWorldExternalApplication
  /// <summary>
  /// A minimal external application saying hello.
  /// Explain the concept of an external application and its minimal interface.
  /// Explore how to create a manifest file for an external application.
  /// </summary>
  [Regeneration( RegenerationOption.Manual )]
  public class Lab6_1_HelloWorldExternalApplication : IExternalApplication
  {
    /// <summary>
    /// This method will run automatically when Revit start up.
    /// It shows a message box. Initiative task can be done in this function.
    /// </summary>
    public Result OnStartup(
      UIControlledApplication a )
    {
      LabUtils.InfoMsg( "Hello World from an external application in C#." );
      return Result.Succeeded;
    }

    public Result OnShutdown(
      UIControlledApplication a )
    {
      return Result.Succeeded;
    }
  }
  #endregion // Lab6_1_HelloWorldExternalApplication

  #region Lab6_2_Ribbon
  /// <summary>
  /// Add various controls to the Revit ribbon.
  /// </summary>
  [Regeneration( RegenerationOption.Manual )]
  public class Lab6_2_Ribbon : IExternalApplication
  {
    public Result OnStartup(
      UIControlledApplication a )
    {
      try
      {
        CreateRibbonItems( a );
      }
      catch ( Exception ex )
      {
        LabUtils.InfoMsg( ex.Message );
        return Result.Failed;
      }
      return Result.Succeeded;
    }

    public Result OnShutdown(
      UIControlledApplication a )
    {
      return Result.Succeeded;
    }

    void CreateRibbonItems( UIControlledApplication a )
    {
      // get the path of our dll
      string addInPath = GetType().Assembly.Location;
      string imgDir = Path.Combine( Path.GetDirectoryName( addInPath ), "img" );

      const string panelName = "Lab 6 Panel";

      const string cmd1 = "Labs.Lab1_1_HelloWorld";
      const string name1 = "HelloWorld";
      const string text1 = "Hello World";
      const string tooltip1 = "Run Lab1_1_HelloWorld command";
      const string img1 = "ImgHelloWorld.png";
      const string img31 = "ImgHelloWorldSmall.png";

      const string cmd2 = "Labs.Lab1_2_CommandArguments";
      const string name2 = "CommandArguments";
      const string text2 = "Command Arguments";
      const string tooltip2 = "Run Lab1_2_CommandArguments command";
      const string img2 = "ImgCommandArguments.png";
      const string img32 = "ImgCommandArgumentsSmall.png";

      const string name3 = "Lab1Commands";
      const string text3 = "Lab 1 Commands";
      const string tooltip3 = "Run a Lab 1 command";
      const string img33 = "ImgCommandSmall.png";

      // create a Ribbon Panel

      RibbonPanel panel = a.CreateRibbonPanel( panelName );

      // add a button for Lab1's Hello World command

      PushButtonData pbd = new PushButtonData( name1, text1, addInPath, cmd1 );
      PushButton pb1 = panel.AddItem( pbd ) as PushButton;
      pb1.ToolTip = tooltip1;
      pb1.LargeImage = new BitmapImage( new Uri( Path.Combine( imgDir, img1 ) ) );

      // add a vertical separation line in the panel
      panel.AddSeparator();

      // prepare data for creating stackable buttons
      PushButtonData pbd1 = new PushButtonData( name1 + " 2", text1, addInPath, cmd1 );
      pbd1.ToolTip = tooltip1;
      pbd1.Image = new BitmapImage( new Uri( Path.Combine( imgDir, img31 ) ) );

      PushButtonData pbd2 = new PushButtonData( name2, text2, addInPath, cmd2 );
      pbd2.ToolTip = tooltip2;
      pbd2.Image = new BitmapImage( new Uri( Path.Combine( imgDir, img32 ) ) );

      PulldownButtonData pbd3 = new PulldownButtonData( name3, text3 );
      pbd3.ToolTip = tooltip3;
      pbd3.Image = new BitmapImage( new Uri( Path.Combine( imgDir, img33 ) ) );

      // add stackable buttons
      IList<RibbonItem> ribbonItems = panel.AddStackedItems( pbd1, pbd2, pbd3 );

      // add two push buttons as sub-items of the Lab 1 commands
      PulldownButton pb3 = ribbonItems[2] as PulldownButton;

      pbd = new PushButtonData( name1, text1, addInPath, cmd1 );
      PushButton pb3_1 = pb3.AddPushButton( pbd );
      pb3_1.ToolTip = tooltip1;
      pb3_1.LargeImage = new BitmapImage( new Uri( Path.Combine( imgDir, img1 ) ) );

      pbd = new PushButtonData( name2, text2, addInPath, cmd2 );
      PushButton pb3_2 = pb3.AddPushButton( pbd );
      pb3_2.ToolTip = tooltip2;
      pb3_2.LargeImage = new BitmapImage( new Uri( Path.Combine( imgDir, img2 ) ) );
    }
  }
  #endregion // Lab6_2_Ribbon

  #region Lab6_3_PreventSaveEvent
  /// <summary>
  /// This external application subscribes to, handles,
  /// and unsubscribes from the document saving event.
  /// Using the event mechnism pre-event features, some actions can be prevented.
  /// This sample displays a message dialog to let the user decide whether to save changes to the document.
  /// Note: The document must have been already saved to disk prior to running this application.
  /// </summary>
  [Regeneration( RegenerationOption.Manual )]
  public class Lab6_3_PreventSaveEvent : IExternalApplication
  {
    public Result OnStartup(
      UIControlledApplication a )
    {
      try
      {
        // subscribe to the DocumentSaving event:
        a.ControlledApplication.DocumentSaving
          += new EventHandler<DocumentSavingEventArgs>(
            a_DocumentSaving );
      }
      catch ( Exception ex )
      {
        LabUtils.InfoMsg( ex.Message );
        return Result.Failed;
      }
      return Result.Succeeded;
    }

    public Result OnShutdown(
      UIControlledApplication a )
    {
      // remove the event subscription:
      a.ControlledApplication.DocumentSaving
        -= new EventHandler<DocumentSavingEventArgs>(
          a_DocumentSaving );
      return Result.Succeeded;
    }

    // Show a message to decide whether to save the document.
    void a_DocumentSaving(
      object obj,
      DocumentSavingEventArgs args )
    {
      // Ask whether to prevent from saving:

      args.Cancel = args.Cancellable
        && LabUtils.QuestionMsg(
          "Saving event handler was triggered.\r\n"
          + "Using the pre-event mechanism, we can cancel the save.\r\n"
          + "Continue saving the document?" );
    }
  }
  #endregion //Lab6_3_PreventSaveEvent

  #region Lab6_4_DismissDialog
  /// <summary>
  /// This external application subscribes to, handles,
  /// and unsubscribe from the dialog box showing event.
  /// Its can dismiss the "Family already exists" dialog when
  /// reloading an already loaded family, and overwrite the existing version.
  /// It dismisses message box by simulating a click on the "Yes" button.
  /// In addition, it can dismiss the property dialog.
  /// </summary>
  [Regeneration( RegenerationOption.Manual )]
  public class Lab6_4_DismissDialog : IExternalApplication
  {
    public Result OnStartup(
      UIControlledApplication a )
    {
      a.DialogBoxShowing
        += new EventHandler<DialogBoxShowingEventArgs>(
          DismissDialog );

      return Result.Succeeded;
    }

    public Result OnShutdown(
      UIControlledApplication a )
    {
      a.DialogBoxShowing
        -= new EventHandler<DialogBoxShowingEventArgs>(
          DismissDialog );

      return Result.Succeeded;
    }

    void DismissDialog(
      object sender,
      DialogBoxShowingEventArgs e )
    {
      TaskDialogShowingEventArgs te = e as TaskDialogShowingEventArgs;
      if ( te != null )
      {
        if ( te.DialogId == "TaskDialog_Family_Already_Exists" )
        {
          // In this task dialog, 1001 maps to the first button,
          // which is the "override the existing version"
          int iReturn = 1001;

          // Set OverrideResult argument to 1001 mimic clicking the first button.
          e.OverrideResult( iReturn );

          // 1002 maps the second button in this dialog.
          // DialogResult.Cancel maps to the cancel button.
        }
      }
      else
      {
        MessageBoxShowingEventArgs msgArgs = e as MessageBoxShowingEventArgs;
        if ( null != msgArgs ) // this is a message box
        {
          e.OverrideResult( (int) WinForms.DialogResult.Yes );
          Debug.Print( "Dialog id is {0}\r\nMessage is {1}",
            msgArgs.HelpId, msgArgs.Message );
        }
        else // this is some other dialog, for example, element property dialog.
        {
          //Use the HelpId to identify the dialog.
          if ( e.HelpId == 1002 ) // Element property dialog's HelpId is 1002
          {
            e.OverrideResult( (int) WinForms.DialogResult.No );
            Debug.Print(
              "We just dismissed the element property dialog "
              + "and set the return value to No." );
          }
        }
      }
    }
  }
  #endregion //Lab6_4_DismissDialog

  #region Lab6_5_RibbonExplorer
  /// <summary>
  /// List the contents of the Revit ribbon in the Visual Studio debug output window.
  /// </summary>
  [Transaction( TransactionMode.ReadOnly )]
  [Regeneration( RegenerationOption.Manual )]
  public class Lab6_5_RibbonExplorer : IExternalCommand
  {
    public Result Execute(
      ExternalCommandData commandData,
      ref string message,
      ElementSet elements )
    {
      UIApplication app = commandData.Application;
      List<RibbonPanel> panels = app.GetRibbonPanels();
      foreach( RibbonPanel panel in panels )
      {
        Debug.Print( panel.Name );
        IList<RibbonItem> items = panel.GetItems();
        foreach ( RibbonItem item in items )
        {
          RibbonItemType t = item.ItemType;
          if ( RibbonItemType.PushButton == t )
          {
            PushButton b = item as PushButton;
            Debug.Print( "  {0} : {1}", item.ItemText, b.Name );
          }
          else
          {
            Debug.Assert( RibbonItemType.PulldownButton == t, "expected pulldown button" );
            PulldownButton b = item as PulldownButton;
            Debug.Print( "  {0} : {1}", item.ItemText, b.Name );
            foreach ( RibbonItem item2 in b.GetItems() )
            {
              Debug.Assert( RibbonItemType.PushButton == item2.ItemType, "expected push button in pulldown menu" );
              Debug.Print( "    {0} : {1}", item2.ItemText, ((PushButton)item2).Name );
            }
          }
        }
      }
      return Result.Failed;
    }
  }
  #endregion // Lab6_5_RibbonExplorer
}

File: snapshot/2011/src/rac/exercise/vb/Labs6.vb
#Region "Header"
'Revit API .NET Labs

'Copyright (C) 2006-2010 by Autodesk, Inc.

'Permission to use, copy, modify, and distribute this software
'for any purpose and without fee is hereby granted, provided
'that the above copyright notice appears in all copies and
'that both that copyright notice and the limited warranty and
'restricted rights notice below appear in all supporting
'documentation.

'AUTODESK PROVIDES THIS PROGRAM "AS IS" AND WITH ALL FAULTS.
'AUTODESK SPECIFICALLY DISCLAIMS ANY IMPLIED WARRANTY OF
'MERCHANTABILITY OR FITNESS FOR A PARTICULAR USE.  AUTODESK, INC.
'DOES NOT WARRANT THAT THE OPERATION OF THE PROGRAM WILL BE
'UNINTERRUPTED OR ERROR FREE.

'Use, duplication, or disclosure by the U.S. Government is subject to
'restrictions set forth in FAR 52.227-19 (Commercial Computer
'Software - Restricted Rights) and DFAR 252.227-7013(c)(1)(ii)
'(Rights in Technical Data and Computer Software), as applicable.
#End Region

#Region "Namespaces"
Imports System
Imports System.Collections.Generic
Imports System.Diagnostics
Imports System.IO
Imports System.Reflection
Imports System.Linq
Imports WinForms = System.Windows.Forms
Imports System.Windows.Media.Imaging
Imports Microsoft.VisualBasic.Constants
' for ribbon, requires references to PresentationCore and WindowsBase .NET assemblies
Imports Autodesk.Revit.ApplicationServices
Imports Autodesk.Revit.Attributes
Imports Autodesk.Revit.DB
Imports Autodesk.Revit.DB.Architecture
Imports Autodesk.Revit.DB.Events
Imports Autodesk.Revit.UI
Imports Autodesk.Revit.UI.Events
#End Region

Namespace Labs

#Region "Lab6_1_HelloWorldExternalApplication"
    ''' <summary>
    ''' A minimal external application saying hello.
    ''' Explain the concept of an external application and its minimal interface.
    ''' Explore how to create a manifest file for an external application.
    ''' </summary>
    <Regeneration(RegenerationOption.Manual)> _
    Public Class Lab6_1_HelloWorldExternalApplication
        Implements IExternalApplication

        ''' <summary>
        ''' This method will run automatically when Revit start up.
        ''' It shows a message box. Initiative task can be done in this function.
        ''' </summary>
        Public Function OnStartup( _
            ByVal a As UIControlledApplication) _
            As Result _
            Implements IExternalApplication.OnStartup

            LabUtils.InfoMsg("Hello World from an external application in C#.")

            Return Result.Succeeded
        End Function

        Public Function OnShutdown( _
        ByVal a As UIControlledApplication) _
        As Result _
        Implements IExternalApplication.OnShutdown

            Return Result.Succeeded

        End Function

    End Class
#End Region

#Region "Lab6_2_Ribbon"
    ''' <summary>
    ''' Add various controls to the Revit ribbon.
    ''' </summary>
    <Regeneration(RegenerationOption.Manual)> _
    Public Class Lab6_2_Ribbon
        Implements IExternalApplication

        Public Function OnShutdown( _
            ByVal a As UIControlledApplication) _
            As Result _
            Implements IExternalApplication.OnShutdown

            Return Result.Succeeded
        End Function

        Public Function OnStartup( _
            ByVal a As UIControlledApplication) _
            As Result _
            Implements IExternalApplication.OnStartup

            Try
                CreateRibbonItems(a)
            Catch ex As Exception
                LabUtils.InfoMsg(ex.Message)
                Return Result.Failed
            End Try

            Return Result.Succeeded
        End Function

        Private Sub CreateRibbonItems(ByVal a As UIControlledApplication)
            ' get the path of our dll
            Dim addInPath As String = [GetType]().Assembly.Location
            Dim imgDir As String = Path.Combine(Path.GetDirectoryName(addInPath), "img")

            Const panelName As String = "Lab 6 Panel"

            Const cmd1 As String = "Labs.Lab1_1_HelloWorld"
            Const name1 As String = "HelloWorld"
            Const text1 As String = "Hello World"
            Const tooltip1 As String = "Run Lab1_1_HelloWorld command"
            Const img1 As String = "ImgHelloWorld.png"
            Const img31 As String = "ImgHelloWorldSmall.png"

            Const cmd2 As String = "Labs.Lab1_2_CommandArguments"
            Const name2 As String = "CommandArguments"
            Const text2 As String = "Command Arguments"
            Const tooltip2 As String = "Run Lab1_2_CommandArguments command"
            Const img2 As String = "ImgCommandArguments.png"
            Const img32 As String = "ImgCommandArgumentsSmall.png"

            Const name3 As String = "Lab1Commands"
            Const text3 As String = "Lab 1 Commands"
            Const tooltip3 As String = "Run a Lab 1 command"
            Const img33 As String = "ImgCommandSmall.png"

            ' create a Ribbon Panel

            Dim panel As RibbonPanel = a.CreateRibbonPanel(panelName)

            ' add a button for Lab1's Hello World command

            Dim pbd As New PushButtonData(name1, text1, addInPath, cmd1)
            Dim pb1 As PushButton = TryCast(panel.AddItem(pbd), PushButton)
            pb1.ToolTip = tooltip1
            pb1.LargeImage = New BitmapImage(New Uri(Path.Combine(imgDir, img1)))

            ' add a vertical separation line in the panel
            panel.AddSeparator()

            ' prepare data for creating stackable buttons
            Dim pbd1 As New PushButtonData(name1 + " 2", text1, addInPath, cmd1)
            pbd1.ToolTip = tooltip1
            pbd1.Image = New BitmapImage(New Uri(Path.Combine(imgDir, img31)))

            Dim pbd2 As New PushButtonData(name2, text2, addInPath, cmd2)
            pbd2.ToolTip = tooltip2
            pbd2.Image = New BitmapImage(New Uri(Path.Combine(imgDir, img32)))

            Dim pbd3 As New PulldownButtonData(name3, text3)
            pbd3.ToolTip = tooltip3
            pbd3.Image = New BitmapImage(New Uri(Path.Combine(imgDir, img33)))

            ' add stackable buttons
            Dim ribbonItems As IList(Of RibbonItem) = panel.AddStackedItems(pbd1, pbd2, pbd3)

            ' add two push buttons as sub-items of the Lab 1 commands
            Dim pb3 As PulldownButton = TryCast(ribbonItems(2), PulldownButton)

            pbd = New PushButtonData(name1, text1, addInPath, cmd1)
            Dim pb3_1 As PushButton = pb3.AddPushButton(pbd)
            pb3_1.ToolTip = tooltip1
            pb3_1.LargeImage = New BitmapImage(New Uri(Path.Combine(imgDir, img1)))

            pbd = New PushButtonData(name2, text2, addInPath, cmd2)
            Dim pb3_2 As PushButton = pb3.AddPushButton(pbd)
            pb3_2.ToolTip = tooltip2
            pb3_2.LargeImage = New BitmapImage(New Uri(Path.Combine(imgDir, img2)))
        End Sub
    End Class
#End Region

#Region "Lab6_3_PreventSaveEvent"
    ''' <summary>
    ''' This external application subscribes to, handles,
    ''' and unsubscribes from the document saving event.
    ''' Using the event mechnism pre-event features, some actions can be prevented.
    ''' This sample displays a message dialog to let the user decide whether to save changes to the document.
    ''' Note: The document must have been already saved to disk prior to running this application.
    ''' </summary>
    <Regeneration(RegenerationOption.Manual)> _
    Public Class Lab6_3_PreventSaveEvent
        Implements IExternalApplication

        Public Function OnShutdown( _
            ByVal a As UIControlledApplication) _
            As Result _
            Implements IExternalApplication.OnShutdown

            ' remove the event subscription:
            RemoveHandler a.ControlledApplication.DocumentSaving, AddressOf a_DocumentSaving

            Return Result.Succeeded
        End Function

        Public Function OnStartup( _
        ByVal a As UIControlledApplication) _
        As Result _
        Implements IExternalApplication.OnStartup

            Try
                ' subscribe to the DocumentSaving event:
                AddHandler a.ControlledApplication.DocumentSaving, AddressOf a_DocumentSaving
            Catch ex As Exception
                LabUtils.InfoMsg(ex.Message)
                Return Result.Failed
            End Try

            Return Result.Succeeded
        End Function

        ' Show a message to decide whether to save the document.
        Private Sub a_DocumentSaving(ByVal obj As Object, ByVal args As DocumentSavingEventArgs)
            ' Ask whether to prevent from saving:
            args.Cancel = args.Cancellable AndAlso _
                LabUtils.QuestionMsg("Saving event handler was triggered." _
                                     + vbCr + vbLf + "Using the pre-event mechanism, we can cancel the save." _
                                     + vbCr + vbLf + "Continue saving the document?")
        End Sub

    End Class
#End Region

#Region "Lab6_4_DismissDialog"
    ''' <summary>
    ''' This external application subscribes to, handles,
    ''' and unsubscribe from the dialog box showing event.
    ''' Its can dismiss the "Family already exists" dialog when
    ''' reloading an already loaded family, and overwrite the existing version.
    ''' It dismisses message box by simulating a click on the "Yes" button.
    ''' In addition, it can dismiss the property dialog.
    ''' </summary>
    <Regeneration(RegenerationOption.Manual)> _
    Public Class Lab6_4_DismissDialog
        Implements IExternalApplication

        Public Function OnShutdown( _
            ByVal a As UIControlledApplication) _
            As Result _
            Implements IExternalApplication.OnShutdown

            RemoveHandler a.DialogBoxShowing, AddressOf DismissDialog

            Return Result.Succeeded
        End Function

        Public Function OnStartup( _
            ByVal a As UIControlledApplication) _
            As Result _
            Implements IExternalApplication.OnStartup

            AddHandler a.DialogBoxShowing, AddressOf DismissDialog

            Return Result.Succeeded
        End Function

        Private Sub DismissDialog( _
            ByVal sender As Object, _
            ByVal e As DialogBoxShowingEventArgs)

            Dim te As TaskDialogShowingEventArgs = TryCast(e, TaskDialogShowingEventArgs)
            If te IsNot Nothing Then
                If te.DialogId = "TaskDialog_Family_Already_Exists" Then
                    ' In this task dialog, 1001 maps to the first button,
                    ' which is the "override the existing version"
                    Dim iReturn As Integer = 1001

                    ' Set OverrideResult argument to 1001 mimic clicking the first button.

                    ' 1002 maps the second button in this dialog.
                    ' DialogResult.Cancel maps to the cancel button.
                    e.OverrideResult(iReturn)
                End If
            Else
                Dim msgArgs As MessageBoxShowingEventArgs = TryCast(e, MessageBoxShowingEventArgs)
                If msgArgs IsNot Nothing Then
                    ' this is a message box
                    e.OverrideResult(CInt(WinForms.DialogResult.Yes))

                    Debug.Print( _
                        "Dialog id is {0}" + vbCr + vbLf + "Message is {1}", _
                        msgArgs.HelpId, msgArgs.Message)
                Else
                    ' this is some other dialog, for example, element property dialog.
                    'Use the HelpId to identify the dialog.
                    If e.HelpId = 1002 Then
                        ' Element property dialog's HelpId is 1002
                        e.OverrideResult(CInt(WinForms.DialogResult.No))
                        Debug.Print("We just dismissed the element property dialog " _
                                    + "and set the return value to No.")
                    End If
                End If
            End If
        End Sub


    End Class
#End Region

#Region "Lab6_5_RibbonExplorer"
    ''' <summary>
    ''' List the contents of the Revit ribbon in the Visual Studio debug output window.
    ''' </summary>
    <Transaction(TransactionMode.Automatic)> _
    <Regeneration(RegenerationOption.Manual)> _
    Public Class Lab6_5_RibbonExplorer
        Implements IExternalCommand

        Public Function Execute( _
            ByVal commandData As ExternalCommandData, _
            ByRef message As String, _
            ByVal elements As ElementSet) _
            As Result _
            Implements IExternalCommand.Execute

            Dim app As UIApplication = commandData.Application
            Dim panels As List(Of RibbonPanel) = app.GetRibbonPanels()
            For Each panel As RibbonPanel In panels
                Debug.Print(panel.Name)
                Dim items As IList(Of RibbonItem) = panel.GetItems()
                For Each item As RibbonItem In items
                    Dim t As RibbonItemType = item.ItemType
                    If RibbonItemType.PushButton = t Then
                        Dim b As PushButton = TryCast(item, PushButton)
                        Debug.Print(" {0} : {1}", item.ItemText, b.Name)
                    Else
                        Debug.Assert(RibbonItemType.PulldownButton = t, "expected pulldown button")
                        Dim b As PulldownButton = TryCast(item, PulldownButton)
                        Debug.Print(" {0} : {1}", item.ItemText, b.Name)
                        For Each item2 As RibbonItem In b.GetItems()
                            Debug.Assert(RibbonItemType.PushButton = item2.ItemType, "expected push button in pulldown menu")
                            Debug.Print(" {0} : {1}", item2.ItemText, DirectCast(item2, PushButton).Name)
                        Next
                    End If
                Next
            Next
            Return Result.Failed
        End Function
    End Class
#End Region

End Namespace

